cpp/oneapi/dal/backend/primitives/ndarray.hpp

```diff
@@ -0,0 +1,405 @@
+                stride *= shape[i];
+            }
+        }
+        else if constexpr (order == ndorder::f) {
+            for (std::int64_t i = 0; i < axis_count; i++) {
+                strides[i] = stride;
+                stride *= shape[i];
+            }
+        }
+
+        return strides;
+    }
+
+    void check_if_strides_are_default() const {
+#ifdef ONEDAL_ENABLE_ASSERT
+        const auto default_strides = get_default_strides(this->get_shape());
+        for (std::int64_t i = 0; i < axis_count; i++) {
+            ONEDAL_ASSERT(this->get_strides()[i] == default_strides[i],
+                          "Operation can be applied only to the array with default strides");
+        }
+#endif
+    }
+
+private:
+    ndshape<axis_count> shape_;
+    ndshape<axis_count> strides_;
+};
+
+template <typename T, std::int64_t axis_count, ndorder order = ndorder::c>
+class ndview : public ndarray_base<axis_count, order> {
+    template <typename, std::int64_t, ndorder>
+    friend class ndview;
+
+public:
+    using base = ndarray_base<axis_count, order>;
+    using shape_t = ndshape<axis_count>;
+
+    ndview() : data_(nullptr) {}
+
+    static ndview wrap(T* data, const shape_t& shape) {
+        return ndview{ data, shape };
+    }
+
+    static ndview wrap(T* data, const shape_t& shape, const shape_t& strides) {
+        return ndview{ data, shape, strides };
+    }
+
+    T* get_data() const {
+        return data_;
+    }
+
+    bool has_data() const {
+        return this->get_count() > 0;
+    }
+
+    auto t() const {
+        using tranposed_ndview_t = ndview<T, axis_count, transposed_ndorder_v<order>>;
+        const auto& shape = this->get_shape();
+        const auto& strides = this->get_strides();
+        return tranposed_ndview_t{ data_, shape.t(), strides.t() };
+    }
+
+    template <std::int64_t new_axis_count>
+    auto reshape(const ndshape<new_axis_count>& new_shape) const {
+        using reshaped_ndview_t = ndview<T, new_axis_count, order>;
+        check_reshape_conditions(new_shape);
+        return reshaped_ndview_t{ data_, new_shape };
+    }
+
+#ifdef ONEDAL_DATA_PARALLEL
+    sycl::event prefetch(sycl::queue& queue) const {
+        return queue.prefetch(data_, this->get_count());
+    }
+#endif
+
+protected:
+    explicit ndview(T* data, const shape_t& shape, const shape_t& strides)
+            : base(shape, strides),
+              data_(data) {}
+
+    explicit ndview(T* data, const shape_t& shape) : base(shape), data_(data) {}
+
+    template <std::int64_t new_axis_count>
+    void check_reshape_conditions(const ndshape<new_axis_count>& new_shape) const {
+        ONEDAL_ASSERT(new_shape.get_count() == this->get_count(),
+                      "Total element count must remain unchanged");
+        base::check_if_strides_are_default();
+    }
+
+private:
+    T* data_;
+};
+
+template <typename T, std::int64_t axis_count, ndorder order = ndorder::c>
+class ndarray : public ndview<T, axis_count, order> {
```

Comment:
If every `ndarray` is `ndview`, why we need them both? Ownership strategy can be handled via deleters.

Reply:
Ndview is more lightweight, no need to store additional deleters and count references. As we know, reference counting could affect performance significantly.

I keep in mind the following scenario: we loop over some large array and handle each block. It's handy to get the new block using get_view with offset. If get_view returns array, at each iteration we would need to call sptr constructor, increment the counter and than call destructor, which causes virtual call.